tracer/buffer.go

```diff
@@ -1,60 +1,115 @@
 package tracer
 
 import (
-	"math/rand"
+	"fmt"
 	"sync"
 )
 
 const (
-	spanBufferDefaultMaxSize = 10000
+	// traceBufferDefaultMaxSize is the maximum number of spans we keep in memory.
+	// This is to avoid memory leaks, if above that value, spans are randomly
+	// dropped and ignore, resulting in corrupted tracing data, but ensuring
+	// original program continues to work as expected.
+	traceBufferDefaultMaxSize = 10000
 )
 
-// spansBuffer is a threadsafe buffer for spans.
-type spansBuffer struct {
-	lock    sync.Mutex
+type traceBuffer struct {
+	// spans is a traceBuffer containing all the spans for this trace.
+	// The reason we don't use a channel here, is we regularly need
+	// to walk the array to find out if it's done or not.
 	spans   []*Span
 	maxSize int
-}
 
-func newSpansBuffer(maxSize int) *spansBuffer {
+	traceChan chan<- []*Span
+	errChan   chan<- error
+
+	sync.RWMutex
+}
 
-	// small sanity check on the max size.
+func newTraceBuffer(traceChan chan<- []*Span, errChan chan<- error, maxSize int) *traceBuffer {
 	if maxSize <= 0 {
-		maxSize = spanBufferDefaultMaxSize
+		maxSize = traceBufferDefaultMaxSize
+	}
+	return &traceBuffer{
+		traceChan: traceChan,
+		errChan:   errChan,
+		maxSize:   maxSize,
+	}
+}
+
+func (tb *traceBuffer) doPush(span *Span) {
+	tb.Lock()
+	defer tb.Unlock()
+
+	// if traceBuffer is full, forget span
+	if len(tb.spans) >= tb.maxSize {
+		select {
+		case tb.errChan <- fmt.Errorf("[TODO:christian] exceed traceBuffer size"):
+		default: // if channel is full, drop & ignore error, better do this than stall program
+		}
+		return
+	}
+	// if there's a trace ID mismatch, ignore span
+	if len(tb.spans) > 0 && tb.spans[0].TraceID != span.TraceID {
+		select {
+		case tb.errChan <- fmt.Errorf("[TODO:christian] trace ID mismatch"):
+		default: // if channel is full, drop & ignore error, better do this than stall program
+		}
+		return
 	}
 
-	return &spansBuffer{maxSize: maxSize}
+	tb.spans = append(tb.spans, span)
```

Comment:
This array behaves an awful lot like a buffered channel. Switching to one would remove the need for an explicit lock and with Manu's suggestion of adding a counter it would still be possible to count the number of finished spans.